cluster_experiments/experiment_analysis.py

```diff
@@ -301,6 +302,112 @@ def from_config(cls, config):
         )
 
 
+class PairedTTestClusteredAnalysis(ExperimentAnalysis):
+    """
+    Class to run paired T-test analysis on aggregated data
+
+    Arguments:
+        cluster_cols: list of columns to use as clusters
+        target_col: name of the column containing the variable to measure
+        treatment_col: name of the column containing the treatment variable
+        treatment: name of the treatment to use as the treated group
+        strata_cols: list of index columns for paired t test. Should be a subset or equal to cluster_cols
+
+    Usage:
+
+    ```python
+    from cluster_experiments.experiment_analysis import PairedTTestClusteredAnalysis
+    from scipy.stats import  ttest_rel
+    import pandas as pd
+
+    df = pd.DataFrame({
+        'x': [1, 2, 3, 4, 0, 0, 1, 1],
+        'treatment': ["A", "B", "A", "B"] * 2,
+        'cluster': [1, 2, 3, 4, 1, 2, 3, 4],
+    })
+
+    PairedTTestClusteredAnalysis(
+        cluster_cols=df['cluster'],
+        strata=df['cluster'],
+        target_col=df['x'],
+    ).get_pvalue(df)
+    ```
+    """
+
+    def __init__(
+        self,
+        cluster_cols: List[str],
+        strata_cols: List[str],
+        target_col: str = "target",
+        treatment_col: str = "treatment",
+        treatment: str = "B",
+    ):
+        self.strata_cols = cluster_cols if strata_cols is None else strata_cols
+        self.target_col = target_col
+        self.treatment = treatment
+        self.treatment_col = treatment_col
+        self.cluster_cols = cluster_cols
+
+    def preprocessing(self, df):
+        df_grouped = df.groupby(
+            self.cluster_cols + [self.treatment_col], as_index=False
+        )[self.target_col].mean()
+
+        n_control = df_grouped[self.treatment_col].value_counts()[0]
+        n_treatment = df_grouped[self.treatment_col].value_counts()[1]
+
+        if n_control != n_treatment:
+            logging.warning(
+                f"groups don't have same number of observations, {n_treatment =} and  {n_treatment =}"
+            )
+
+        #         set(self.strata_cols if len ) <= set(self.cluster_cols)
+        assert all(
+            [x in self.cluster_cols for x in self.strata_cols]
+        ), f"strata should be a subset or equal to cluster_cols ({self.cluster_cols = }, {self.strata_cols = })"
+
+        df_pivot = df_grouped.pivot_table(
+            columns=self.treatment_col,
+            index=self.strata_cols,
+            values=self.target_col,
+        )
+
+        if df_pivot.isna().sum().sum() == 0:
```

Comment:
what's this case doing?

Reply:
removing for example cluster 63 from analysis

![image](https://user-images.githubusercontent.com/33734080/217018534-623c6bbc-8db0-47d2-8178-21c23a6e2ba8.png)

Reply:
alright, could you comment this somehow? that his is to handle cases where the pairing doesn't exist